### DISCUSSION TITLE
After upgrade to 2.13.24 Filter Throws On Null Exception

### DISCUSSION THREAD

**Discussion by madsem** [filamentphp/filament#3073] [General]
I just upgraded Filament Forms & Table Builder and now one of my filters is throwing an exception, and I'm unsure if this is a bug or if I've made a whoopsie :)

Filter in `getTableFilters()`:
```php
Filter::make('has_spent')
                ->label('With Spend Only')
                ->default()
                ->query(function (Builder $q) {

                    return $q->whereHas('reports', function ($q) {
                        $q->hasSpend()
                            ->whereDate('date', '>=', $this->fromDate)
                            ->whereDate('date', '<=', $this->toDate);
                    });
                })
                ->toggle(),
```

`Call to a member function reports() on null`

If I call the same View with querystring set, it works. 

I.E, going to http://localhost/accounts throws exception, while http://localhost/accounts?tableFilters[date][date_from]=2022-07-08+09%3A24%3A17&tableFilters[date][date_until]=2022-07-08+09%3A24%3A17&tableFilters[has_spent][isActive]=0
works as expected.

**Comment by danharrin** [comment]
Which specific version broke it? Please downgrade gradually until it works again.

Also, please share the Flare exception URL.

**Comment by madsem** [comment]
Sorry, only mentioned the version in title, here a bit clearer:

Upgraded to `2.13.24` from (I believe) `2.12.19`

Flare:
https://flareapp.io/share/A7EpEpkP#context-context-versions

**Reply by danharrin** [reply]
So please downgrade a patch version at a time until it works

**Reply by danharrin** [reply]
Ohhh I know what it is. The `$q` variable needs to be `$query`, as per the docs.

Every callback parameter name needs to be the same as the docs, since we use injection.

**Reply by madsem** [reply]
wow! I never would have found that LOL.
Thank you @danharrin 

I wonder though, I have other filters, also just using `$q`, those do not throw any exceptions:

```php
protected function getTableFilters(): array
    {
        return [

            Filter::make('date')
                ->form([
                    Grid::make()
                        ->schema([
                            Forms\Components\DatePicker::make('date_from')->default(now()),
                            Forms\Components\DatePicker::make('date_until')->default(now()),
                        ]),
                ]),

            SelectFilter::make('source')
                ->label('Traffic Source')
                ->options(Account::getAccountTypes())
                ->searchable()
                ->query(function (Builder $q, array $data) {
                    $q->when(
                        $data['value'],
                        fn(Builder $q, $source): Builder => $q->where('type', $source)
                    );
                }),

            SelectFilter::make('users')
                ->label('Vikings')
                ->options(User::pluck('name', 'id'))
                ->searchable()
                ->visible(\Auth::user()->hasAnyRoles([
                    'super_admin',
                    'partner',
                ]))
                ->query(function (Builder $q, array $data): Builder {
                    return $q
                        ->when(
                            $data['value'],
                            function (Builder $q, $user) {
                                return $q->where('user_id', $user);
                            }
                        );
                }),

            Filter::make('has_spent')
                ->label('With Spend Only')
                ->default()
                ->query(function (Builder $query) {

                    return $query->whereHas('reports', function ($query) {
                        $query->hasSpend()
                            ->whereDate('date', '>=', $this->fromDate)
                            ->whereDate('date', '<=', $this->toDate);
                    });
                })
                ->toggle(),
        ];
    }
```

I've changed it on the `has_spent` filter, and it indeed works... But the others are working as well?!

**Reply by danharrin** [reply]
Sure they're actually affecting the table contents? I doubt they are.

When you inject `$q`, Laravel will try to resolve it from the container, which just instantiates a new builder. The reason you had an exception is because that new builder doesnt have a model instance associated, so the relationship broke.

**Comment by madsem** [comment]
Ooook, makes sense, basically because it `was` instantiated before using `$query` yes?
That makes sense, thank you :)

And yes, the filters affect the table content but you're right. Looking at it, the `has_spent` is the only one set as `default` filter, I bet when I toggle the others I will run into the same issue.
